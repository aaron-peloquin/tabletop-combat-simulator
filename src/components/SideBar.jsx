import {
  Drawer,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  Typography,
  IconButton,
} from "@material-ui/core/";
import Link from "next/link";

import { connect } from "react-redux";

import ToggleSidebar from "./../store/dispatcher/ToggleSidebar";

const SideBar = (props) => {
  const { dispatch, open } = props
  return <Drawer open={open} anchor="left" onOpen={()=>{ return ToggleSidebar(dispatch); }} onClose={()=>{ return ToggleSidebar(dispatch); }}>
    <Link href="/about" onClick={()=>{ return ToggleSidebar(dispatch); }}><a>About</a></Link>
  </Drawer>;
}

const mapStateToProps = (state) => {
  console.log("ALERT",state);
  return {open:state.sideBar};
}

export default connect(mapStateToProps)(SideBar);