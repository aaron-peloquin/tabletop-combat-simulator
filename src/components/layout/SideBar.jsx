import {SwipeableDrawer, List} from "@material-ui/core/"
import {withStyles} from "@material-ui/core/styles"
import {Home, Info, ViewList} from "@material-ui/icons"
import {connect} from "react-redux"

import toggleSidebar from "./../../store/action/ToggleSidebar"
import SideBarLink from "./SideBarLink"

/**
 * Swipable drawer for navigation
 * @param {obj} props
 *  @param {func} dispatch redux dispatcher
 *  @param {obj} classes is the wrapping MUI styles for all pages
 *  @param {bool} open the Redux state of the sideBar
 *  @param {func} funcToggleSidebar is the redux action (dispatcher) for toggling the sideBar open/closed
 * @return {jsx} Component
 */
const SideBar = (props) => {
  const {dispatch, classes, open, funcToggleSidebar} = props
  return <SwipeableDrawer onOpen={()=>{
    funcToggleSidebar(dispatch)
  }} open={open} anchor="left" onClose={()=>{
    funcToggleSidebar(dispatch)
  }}>
    <div
      className={classes.drawerContainer}
      role="button"
      onClick={()=>{
        funcToggleSidebar(dispatch)
      }}
      onKeyDown={()=>{
        funcToggleSidebar(dispatch)
      }}
    >
      <List>
        <SideBarLink text="Home" url="/"><Home /></SideBarLink>
        <SideBarLink text="About" url="/about"><Info /></SideBarLink>
        <SideBarLink text="Creatures" url="/creatures"><ViewList /></SideBarLink>
        <SideBarLink text="Combat Simulator" url="/combat-simulator"></SideBarLink>
      </List>
    </div>
  </SwipeableDrawer>
}

SideBar.defaultProps = {
  funcToggleSidebar: toggleSidebar,
}

export const mapStateToProps = (state) => {
  return {open: state.sideBar}
}

/** MUI Styles for <SideBar /> */
const styles = {
  drawerContainer: {width: 250},
}

/** Add MUI styles to this component */
const SideBarWithStyles = withStyles(styles)(SideBar)
/** Connect this component to Redux */
export default connect(mapStateToProps)(SideBarWithStyles)
