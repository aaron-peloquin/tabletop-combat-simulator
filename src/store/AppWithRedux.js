import React from 'react'
import {initializeStore} from './store'

const isServer = typeof window === 'undefined'
const reduxStoreSymbol = Symbol("reduxStore")

function getOrCreateStore (initialState) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return initializeStore(initialState)
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[reduxStoreSymbol]) {
    window[reduxStoreSymbol] = initializeStore(initialState)
  }
  return window[reduxStoreSymbol]
}

export default (App) => {
  return class AppWithRedux extends React.Component {
    static async getInitialProps (appContext) {
      // Get or Create the store with `undefined` as initialState
      // This allows you to set a custom default initialState
      const reduxStore = getOrCreateStore()

      // Provide the store to getInitialProps of pages
      appContext.ctx.reduxStore = reduxStore

      let appProps = {}
      if (typeof App.getInitialProps === 'function') {
        appProps = await App.getInitialProps(appContext)
      }

      return {
        ...appProps,
        initialReduxState: reduxStore.getState()
      }
    }

    constructor (props) {
      super(props)
      this.reduxStore = getOrCreateStore(props.initialReduxState)
    }

    render () {
      return <App {...this.props} reduxStore={this.reduxStore} />
    }
  }
}
