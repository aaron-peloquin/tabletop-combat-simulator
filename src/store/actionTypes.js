/**
 * This is an object of Redux Actions. It mostly exists to
 *  ensure we do not create duplicate actions, but it can
 *  also be used as a list of our existing actions
*/
const actions = {
  CopyCreature: 'CopyCreature',
  DeleteAllCreatures: 'DeleteAllCreatures',
  DeleteCreature: 'DeleteCreature',
  SaveCreature: 'SaveCreature',
  SetEditCreature: 'SetEditCreature',
  ToggleSidebar: 'ToggleSidebar',
  RunSimulation: 'RunSimulation',
  UpdateEditCreature: 'UpdateEditCreature',
}

export default actions
