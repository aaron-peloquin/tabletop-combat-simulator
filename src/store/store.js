import {createStore, applyMiddleware, combineReducers} from "redux"
import {composeWithDevTools} from "redux-devtools-extension"
import thunkMiddleware from "redux-thunk"

/** Reducers */
import combat from "./reducer/combat"
import creatures from "./reducer/creatures"
import sideBar from "./reducer/sideBar"
import editing from "./reducer/editing"

/** Reducers */
const reducers = combineReducers({
  combat,
  creatures,
  editing,
  sideBar,
})

/**
 * Initializes the store
 * @param {obj} initialState the state our app starts with
 * @return {obj} the redux store
 */
export function initializeStore(initialState = initialState) {
  return createStore(reducers, initialState, composeWithDevTools(applyMiddleware(thunkMiddleware)))
}
