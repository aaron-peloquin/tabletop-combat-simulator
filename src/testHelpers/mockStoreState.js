export default {
  "creatures": [
    {
      "name": "Goplin",
      "cr": "1/4",
      "hash": "hivn3p0k",
      "description": "A stout monster",
      "defaultInitiative": 10,
      "armor": 11,
      "hp": 5,
      "hitDiceEquation": "1d20+1",
      "damageDiceEquation": "1d4"
    },
    {
      "name": "Elven Ranger",
      "cr": "1/2",
      "hash": "h3dxf5wl",
      "description": "A swift hunter",
      "defaultInitiative": 16,
      "armor": 13,
      "hp": 10,
      "hitDiceEquation": "1d20+8",
      "damageDiceEquation": "1d6+3"
    },
    {
      "name": "Knight",
      "cr": "1",
      "hash": "hee9p286",
      "description": "An honorable soldier",
      "defaultInitiative": 11,
      "armor": 18,
      "hp": 25,
      "hitDiceEquation": "1d20+4",
      "damageDiceEquation": "1d12+4"
    },
    {
      "name": "Lizardfolk",
      "cr": "3",
      "hash": "howx26rt",
      "description": "An ancient sneaky monster",
      "defaultInitiative": 14,
      "armor": 14,
      "hp": 15,
      "hitDiceEquation": "1d20+5",
      "damageDiceEquation": "1d10+2"
    }
  ],
  "sideBar": false,
}