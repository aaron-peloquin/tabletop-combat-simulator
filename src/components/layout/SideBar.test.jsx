import { Provider } from "react-redux"
import { createShallow } from "@material-ui/core/test-utils"
import toJson from "enzyme-to-json"
import { initializeStore } from "../../store/store"

import SideBar, { mapStateToProps } from "./SideBar"
import mockStoreState from "./../../testHelpers/mockStoreState"

describe("<SideBar />", ()=>{
  const RenderShallowUntilComponent = createShallow({"untilSelector":"SideBar"})
  let store, props, Component
  beforeEach(async ()=>{
    store = initializeStore(mockStoreState)
    props = mapStateToProps(store.getState())

    /** Attach spy to store.dispatch */
    store.dispatch = jest.fn(store.dispatch)
    Component = RenderShallowUntilComponent(<Provider store={store}><SideBar {...props} /></Provider>)
  })

  it("loads", async () => {
    expect(typeof Component).toBe("object")
  })

  it("is closed by default", () => {
    expect(props.open).toBe(false)
  })

  /**
   * Simulate clicking the wrapping <div /> twice,
   * should call dispatcher twice and should toggle from false to true,
   * then back to false
   */
  it("is properly toggles", () => {
    const wrappingDiv = Component.find("div")
    wrappingDiv.simulate("click")
    expect(store.dispatch).toBeCalledTimes(1)
    props = mapStateToProps(store.getState())
    expect(props.open).toBe(true)
    wrappingDiv.simulate("click")
    expect(store.dispatch).toBeCalledTimes(2)
    props = mapStateToProps(store.getState())
    expect(props.open).toBe(false)
  })

  it("contains a <List />", () => {
    expect(Component.find("WithStyles(List)").length).toBe(1)
  })

  it("contains at least 3 <SideBarLink />", () => {
    expect(Component.find("WithStyles(SideBarLink)").length).toBeGreaterThanOrEqual(3)
  })

  it("snapshots", () => {
    expect(toJson(Component)).toMatchSnapshot()
  })
})