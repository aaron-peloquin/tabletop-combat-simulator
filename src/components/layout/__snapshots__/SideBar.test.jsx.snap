// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideBar /> snapshots 1`] = `
<WithTheme(SwipeableDrawer)
  anchor="left"
  onClose={[Function]}
  onOpen={[Function]}
  open={false}
>
  <div
    className="SideBar-drawerContainer-1"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
  >
    <WithStyles(List)>
      <WithStyles(SideBarLink)
        text="Home"
        url="/"
      >
        <pure(HomeIcon) />
      </WithStyles(SideBarLink)>
      <WithStyles(SideBarLink)
        text="About"
        url="/about"
      >
        <pure(InfoIcon) />
      </WithStyles(SideBarLink)>
    </WithStyles(List)>
  </div>
</WithTheme(SwipeableDrawer)>
`;
